### PR TITLE
Enable Auto Port Selection for testing

### LINKED ISSUE #125: Dynamically generate test server port.
For unit testing APIs, run Server in test mode with automatically generated port number to avoid port conflicts.  This port number should be publicly exposed so that tests can build and endpoint with this port.  See PR #123

### PR DESCRIPTION
**Pull request by andyday**
This fixes #125

### CODE REVIEW

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 1]
```diff
@@ -40,18 +40,21 @@
...
 @Slf4j
+@Accessors(fluent = true)
 public class Server implements Routes {
   private final XConfig config;
   private final Tls tls;
   private final XrpcConnectionContext.Builder contextBuilder;
   private final MetricRegistry metricRegistry = new MetricRegistry();
   @Getter private final RouteBuilder routeBuilder = new RouteBuilder();
 
+  @Getter private int port;
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: I think it's confusing that this doesn't match the configured port under some conditions.

What about having a `getActualPort` method that gets the port dynamically from the channel?

Alternately, documenting when this gets initialized and renaming to `actualPort` would be fine.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 2]
```diff
@@ -40,18 +40,21 @@
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import lombok.Getter;
+import lombok.experimental.Accessors;
 import lombok.extern.slf4j.Slf4j;
 import org.slf4j.LoggerFactory;
 
 /** An xprc server. */
 @Slf4j
+@Accessors(fluent = true)
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: I don't think we should add this here without adding it consistently across the API (like in `XrpcRequest` and other endpoints as well).

I'm fine with whatever, but we should try to have our public API be consistent.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java (hunk 1 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: I've added javadocs to clarify this.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java (hunk 2 above)
Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: leaving this as fluent for now with the plan to change all to fluent in another pr.

Comment (jkinkead) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: Sounds great!

Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: see #142

**Comment by andyday** [comment]
@jkinkead ready for re-review

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 3]
```diff
@@ -40,18 +40,26 @@
...
 public class Server implements Routes {
   private final XConfig config;
   private final Tls tls;
   private final XrpcConnectionContext.Builder contextBuilder;
   private final MetricRegistry metricRegistry = new MetricRegistry();
   @Getter private final RouteBuilder routeBuilder = new RouteBuilder();
 
+  /**
+   * Actual port on which this server is running. This will differ from the configured port of it is
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: I really think the name should reflect this (`actualPort`). You should also be clear when this gets initialized - and may also want to use an obviously-bad placeholder, like `-1`, as an initial value.

Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: the main use of this is in testing.  all the tests need to no is which port the server is running on.  to add the word actual adds to the confusion in my view

**Comment by andyday** [comment]
think i've got the port stuff squared.  let me know.

File: demos/people/src/main/java/com/nordstrom/xrpc/demos/people/Application.java [hunk 4]
```diff
@@ -25,18 +25,12 @@
...
 @Slf4j
 public class Application {
-  private final Server server;
-
-  public Application(Config config) {
-    // Build your server. This overrides the default configuration with values from
-    // src/main/resources/demo.conf.
-    this.server = new Server(config);
-
+  public static void configure(Server server) {
```
Comment (jkinkead) [inline demos/people/src/main/java/com/nordstrom/xrpc/demos/people/Application.java]: This seems like a nice pattern.

File: xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java [hunk 5]
```diff
@@ -222,6 +253,13 @@ public void listenAndServe() throws IOException {
     }
 
     channel = future.channel();
+    InetSocketAddress actualAddress = (InetSocketAddress) channel.localAddress();
+    log.info("Listening at {}", actualAddress.getAddress().getCanonicalHostName());
+  }
+
+  public String localEndpoint() {
```
Comment (jkinkead) [inline xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: Maybe add a `@throws NullPointerException if called before listenAndServe`?

Comment (andyday) [reply xrpc/src/main/java/com/nordstrom/xrpc/server/Server.java]: i'll add that on the next pr.

**Pull request merged**